daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/SAXInfosetInputter.scala

```diff
@@ -0,0 +1,157 @@
+
+import scala.util.Try
+
+import org.apache.daffodil.api.DFDL
+import org.apache.daffodil.api.DFDL.DaffodilInputContentHandler
+import org.apache.daffodil.api.DFDL.DaffodilUnhandledSAXException
+import org.apache.daffodil.api.DFDL.DaffodilUnparseErrorSAXException
+import org.apache.daffodil.dpath.NodeInfo
+import org.apache.daffodil.infoset.InfosetInputterEventType.EndDocument
+import org.apache.daffodil.util.Maybe.One
+import org.apache.daffodil.util.MaybeBoolean
+import org.apache.daffodil.util.Misc
+import org.apache.daffodil.xml.XMLUtils
+
+class SAXInfosetInputter(ch: DaffodilInputContentHandler, dp: DFDL.DataProcessor, output: DFDL
+.Output) extends InfosetInputter with DFDL.ConsumerCoroutine {
+
+  var isDone = false
+  var currentEvent: DFDL.SaxInfosetEvent = _
+  var nextEvent: DFDL.SaxInfosetEvent = _
+  val contentHandlerEvent: DFDL.SaxInfosetEvent = new DFDL.SaxInfosetEvent
+
+  override def getEventType(): InfosetInputterEventType = currentEvent.eventType.orNull
+
+  override def getLocalName(): String = currentEvent.localName.orNull
+
+  override def getNamespaceURI(): String = currentEvent.namespaceURI.orNull
+
+  override def getSimpleText(primType: NodeInfo.Kind): String = {
+    val res = if (currentEvent.simpleText.isDefined) {
+      currentEvent.simpleText.get
+    } else (
+      throw new NonTextFoundInSimpleContentException(getLocalName())
+    )
+    if (primType.isInstanceOf[NodeInfo.AnyURI.Kind] && res.nonEmpty) {
+      try {
+        val uri = new URI(res)
+        if (!uri.getPath.startsWith("/")) {
+          // TDML files must allow blob URI's to be relative, but Daffodil
+          // requires them to be absolute with a scheme. So search for the file
+          // using TDML semantics and convert to an absolute URI
+          val abs = Misc.searchResourceOption(uri.getPath, None)
+          abs.get.toString
+        } else {
+          res
+        }
+      } catch {
+        case _: URISyntaxException => res
+      }
+    } else {
+      if (primType.isInstanceOf[NodeInfo.String.Kind]) {
+        val remapped = XMLUtils.remapPUAToXMLIllegalCharacters(res)
+        remapped
+      } else {
+        res
+      }
+    }
+  }
+
+  override def isNilled(): MaybeBoolean = {
+    val _isNilled = if (currentEvent.nilValue.isDefined) {
+      val nilValue = currentEvent.nilValue.get
+      if (nilValue == "true" || nilValue == "1") {
+        MaybeBoolean(true)
+      } else if (nilValue == "false" || nilValue == "0") {
+        MaybeBoolean(false)
+      } else {
+        throw new InvalidInfosetException("xsi:nil property is not a valid boolean: '" + nilValue +
+          "' for element " + getLocalName())
+      }
+    } else {
+      MaybeBoolean.Nope
+    }
+    _isNilled
+  }
+
+  //called without changing any state
+  override def hasNext(): Boolean = {
+    if (isDone) {
+      false
+    } else {
+      val event = this.resume(ch, Try(currentEvent))
+      nextEvent = copyEvent(event.getOrElse(null))
```

Comment:
Yes, the contentHandler has just 1 event that it modifies, which is why I chose to do the copy way. But I updated to have just 1 instance of the 3 events and they we copy between the fields rather than the allocating a new infosetEvent on each copy